File: baekjoon/js/4949.js
const fs = require("fs");
const buf = fs.readFileSync(0, "utf-8").toString().trim();

const inputs = buf.split("\n");

const result = [];

for (const input of inputs) {
  if (input === ".") {
    break;
  }

  const stack = [];
  let isValid = true;

  for (const c of input) {
    if (c === "(" || c === "[") {
      stack.push(c);
    } else if (c === ")") {
      if (stack.length === 0 || stack.pop() !== "(") {
        isValid = false;
        break;
      }
    } else if (c === "]") {
      if (stack.length === 0 || stack.pop() !== "[") {
        isValid = false;
        break;
      }
    }
  }

  result.push(stack.length === 0 && isValid ? "yes" : "no");
}

console.log(result.join("\n"));

File: baekjoon/rust/4949.rs
use std::io::stdin;

fn main() {
    let mut lines = stdin().lines();

    let mut result = Vec::new();

    while let Some(Ok(line)) = lines.next() {
        if line == "." {
            break;
        }

        let input = line.trim();

        let mut stack = Vec::new();
        let mut is_valid = true;

        for c in input.chars() {
            if c == '(' || c == '[' {
                stack.push(c);
            } else if c == ')' {
                let val = stack.pop();

                if val.is_none() || val.unwrap() != '(' {
                    is_valid = false;
                    break;
                }
            } else if c == ']' {
                let val = stack.pop();

                if val.is_none() || val.unwrap() != '[' {
                    is_valid = false;
                    break;
                }
            }
        }

        if stack.is_empty() && is_valid {
            result.push("yes");
        } else {
            result.push("no")
        }
    }

    print!("{}", result.join("\n"));
}
